File: python_solution/medium/Solution1219.py
from typing import List


class Solution:

    def getMaximumGold(self, grid: List[List[int]]) -> int:
        rowLength = len(grid)
        columnLength = len(grid[0])
        sum = 0
        visited = [[False for _ in range(rowLength)] for _ in range(columnLength)]
        for i in range(rowLength):
            for j in range(columnLength):
                sum = max(sum, self.findTheMaxPath(i, j, grid, visited))
        return sum

    def findTheMaxPath(self, i, j, grid, visited):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid) or visited[i][j] or grid[i][j] == 0:
            return 0
        visited[i][j] = True
        sum = grid[i][j]
        sum += max(max(self.findTheMaxPath(i + 1, j, grid, visited), self.findTheMaxPath(i - 1, j, grid, visited)),
                   max(self.findTheMaxPath(i, j + 1, grid, visited), self.findTheMaxPath(i, j - 1, grid, visited)))
        visited[i][j] = False
        return sum


if __name__ == "__main__":
    solution = Solution()
    grid = [[0,6,0],[5,8,7],[0,9,0]]
    sum = solution.getMaximumGold(grid)
    print(sum)

File: Java_solution/medium/Solution1219.java
package medium;

public class Solution1219 {
    public static void main(String[] args) {
        Solution1219 solution1219 = new Solution1219();
        int[][] grid = new int[][]{{0,6,0},{5,8,7},{0,9,0}};
        System.out.println(solution1219.getMaximumGold(grid));
    }
    public int getMaximumGold(int[][] grid) {
        int sum = 0;
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                sum = Math.max(findTheMaxGold(i, j, grid, visited),sum);
            }
        }
        return sum;
    }

    private int findTheMaxGold(int i, int j, int[][] grid, boolean[][] visited) {
        if (i >= grid.length || i < 0 || j >= grid[0].length || j < 0 || visited[i][j] || grid[i][j] == 0) {
            return 0;
        }
        int sum = grid[i][j];
        //每次没有将grid[i][j]的值加入到总和中
        visited[i][j]=true;
        //此处是四个方向都进行了访问，实际题目要求是仅能访问一边
        sum +=Math.max( Math.max( findTheMaxGold(i+1,j,grid,visited),findTheMaxGold(i-1,j,grid,visited)),Math.max( findTheMaxGold(i,j+1,grid,visited),findTheMaxGold(i,j-1,grid,visited)));
        visited[i][j]=false;
        return sum;
        //缺少回溯
    }
}
